infra/terraform/terraform.go

```diff
@@ -64,45 +71,73 @@ func NewFromState(config Config, stateConfig infra.ProvisionerState) (*terraform
 }
 
 func (r *terraform) Create(withInstaller bool) (installer infra.Node, err error) {
-	err = system.CopyFile(filepath.Join(r.stateDir, "terraform.tf"), r.ScriptPath)
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
+	ctx, cancel := context.WithTimeout(context.Background(), terraformTimeout)
+	defer cancel()
 
-	output, err := r.boot()
+	err, nfiles := system.CopyAll(r.ScriptPath, r.stateDir)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
+	if nfiles == 0 {
+		return nil, trace.Errorf("No Terraform configs at %s", r.ScriptPath)
+	}
+
+	// sometimes terraform cannot receive all required params
+	// most often public IPs take time to allocate (on Azure)
+	for {
+		err, retry := r.terraform()
+		if err == nil {
+			if withInstaller {
+				nodes := r.pool.Nodes()
+				if len(nodes) == 0 { // should not happen, and doesn't make sense to retry
+					return nil, trace.Errorf("Zero nodes were allocated")
+				}
+				r.installerIP = nodes[0].Addr()
+				return nodes[0], nil
+			}
+			return nil, nil
+		}
 
-	if withInstaller {
-		// find installer public IP
-		match := reInstallerIP.FindStringSubmatch(output)
-		if len(match) != 2 {
-			return nil, trace.NotFound(
-				"failed to extract installer IP from terraform output: %v", match)
+		if !retry {
+			return nil, trace.Wrap(err, "Terraform creation failed")
 		}
-		r.installerIP = strings.TrimSpace(match[1])
+		select {
+		case <-ctx.Done():
+			return nil, trace.Wrap(err, "Terraform creation timed out")
+		case <-time.After(terraformRepeatAfter):
+			continue
+		}
+	}
+
+}
+
+func (r *terraform) terraform() (err error, retry bool) {
```

Comment:
That's very unusual to have error as a first return argument (unless it's the only one). Can you swap them?